web/source/osk/visualKeyboard.ts

```diff
@@ -962,49 +962,16 @@ namespace com.keyman.osk {
             // Overwrite the previously-computed percent.
             // NB: the 'percent' suffix is historical, units are percent on desktop devices, but pixels on touch devices
             // All key widths and paddings are rounded for uniformity
-            var keyPercent: number, padPercent: number, totalPercent=0;
-            for(j=0; j<keys.length-1; j++) {
-              keyPercent = keys[j]['widthpc'] * objectWidth;
-              keys[j]['widthpc']=keyPercent;
-              padPercent = keys[j]['padpc'] * objectWidth;
-              keys[j]['padpc']=padPercent;
-
-              // Recompute center's x-coord with exact, in-browser values.
-              (<keyboards.ActiveKey> keys[j]).proportionalX = (totalPercent + padPercent + (keyPercent/2))/objectWidth;
-              (<keyboards.ActiveKey> keys[j]).proportionalWidth = keyPercent / objectWidth;
-
-              totalPercent += padPercent+keyPercent;
-            }
-
-            // Allow for right OSK margin (15 layout units)
-            let rightMargin = keyboards.ActiveKey.DEFAULT_RIGHT_MARGIN*objectWidth/layer.totalWidth;
-            totalPercent += rightMargin;
-
-            // If a single key, and padding is negative, add padding to right align the key
-            if(keys.length == 1 && parseInt(keys[0]['pad'],10) < 0) {
-              keyPercent = keys[0]['widthpc'] * objectWidth;
-              keys[0]['widthpc']=keyPercent;
-              totalPercent += keyPercent;
-              keys[0]['padpc']=(objectWidth-totalPercent);
-
-              // Recompute center's x-coord with exact, in-browser values.
-              (<keyboards.ActiveKey> keys[0]).proportionalX = (totalPercent - rightMargin - keyPercent/2)/objectWidth;
-              (<keyboards.ActiveKey> keys[0]).proportionalWidth = keyPercent / objectWidth;
-            } else if(keys.length > 0) {
-              j=keys.length-1;
-              padPercent = keys[j]['padpc'] * objectWidth;
-              keys[j]['padpc']=padPercent;
-              totalPercent += padPercent;
-              keys[j]['widthpc']= keyPercent = (objectWidth-totalPercent);
-
-              // Recompute center's x-coord with exact, in-browser values.
-              (<keyboards.ActiveKey> keys[j]).proportionalX = (objectWidth - rightMargin - keyPercent/2)/objectWidth;
-              (<keyboards.ActiveKey> keys[j]).proportionalWidth = keyPercent / objectWidth;
+            for(j=0; j<keys.length; j++) {
+              key = keys[j];
+              // TODO:  reinstate rounding?
```

Comment:
Can you explain what is missing here?

Reply:
So, it turns out... key width rounding hasn't _actually_ been in place for a little while.  Or, perhaps it kinda has, but with two decimal digits.  (As best as I can tell, that 'precision' is a CSS styling default, not anything we implemented.)  Pretty sure that was unintentional and started in 12.0.

Reply:
Compare this comment's deleted context to the original code from 10.0:

https://github.com/keymanapp/keyman/blob/f16872a99896a2fc208c68ac92a827820ab07e3f/web/source/kmwosk.ts#L2331-L2358

Note that `Math.round` kinda got dropped, never making it in to the `VisualKeyboard` code (that comment's context) where it would have belonged.

Reply:
I have a feeling the rounding was needed for old versions of Android Browser -- or perhaps Opera -- where non-integer values would break. But I haven't looked for evidence!